Re-enable automatic inference of inverse_of for autosave associations

samsondav opened the pull request:
Fixes issues: #9336, #22391, #22358, #23163, #23171, and probably more.

We weren't automatically inferring the inverse_of for autosaved associations. This appears to have been a hack in order to avoid double-save problems with nested attributes but at the expense of not being able to access the parent model during the save callback chain.

Among other things this was preventing validation of parent(s) when creating childen via nested attributes.

The correct fix is to automatically infer inverse_of for autosaved associations and to address the root issue of double queries, which is to ensure that child records saved in callbacks do not cause a duplicate save on the parent.

We do this by wrapping autosaves of associated records in a block which temporarily disables the reverse autosave leading back to the record which triggered it. This ensures we only save each record in the chain once.

Commit:
Re-enable automatic inference of inverse_of for autosave associations

Fixes #9336, fixes #22391, fixes #22358, fixes #23163, fixes #23171, and probably more.

We weren't automatically inferring the inverse_of for autosaved associations. This appears to have been a hack in order to avoid double-save problems with nested attributes but at the expense of not being able to access the parent model during the save callback chain.

Among other things this was preventing validation of parent(s) when creating childen via nested attributes.

The correct fix is to automatically infer inverse_of for autosaved associations and to address the root issue of double queries, which is to ensure that child records saved in callbacks do not cause a duplicate save on the parent.

We do this by autosaves of associated records in a block which temporarily disables reverse autosave on the record which triggered it. This ensures we only save each record in the chain once.

egilburg commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -465,7 +496,11 @@ def save_belongs_to_association(reflection)
-            saved = record.save(:validate => !autosave) if record.new_record? || (autosave && record.changed_for_autosave?)
+            if record.new_record? || (autosave && record.changed_for_autosave?)
+              saved = with_disabled_autosave_on_inverses(reflection.active_record) do
+                record.save(:validate => !autosave)
```
please use 1.9 style hash `(validate: !autosave)`

samphilipd replied:
Done.

egilburg commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -412,6 +414,32 @@ def save_collection_association(reflection)
+        inverse_reflections_with_autosave = Set.new
+
+        klass.reflections.values.each do |reflection|
+          next if reflection.polymorphic? || reflection.inverse_of.try(:options).try(:[], :autosave) != true
```
perhaps:

```ruby
  inverse_of = reflection.inverse_of
  if inverse_of && inverse_of[:options][:autosave] && !reflection.polymorphic?
    inverse_reflections_with_autosave << inverse_of
  end
```

Reduce `try` calls where it's know there can be no nils.

samphilipd replied:
Yup, you are right. Cleaned this up to make it neater.

egilburg commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -412,6 +414,32 @@ def save_collection_association(reflection)
+      # In order to avoid this we temporarily disable autosave on associations
+      # leading back from child records to this parent.
+      def with_disabled_autosave_on_inverses(klass)
+        inverse_reflections_with_autosave = Set.new
```
If you are only using this for linear `each`, what's the reason for using `Set` rather than normal `[]`?

samphilipd replied:
Agreed. Not sure what I was thinking with using Set here.

samphilipd commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -412,6 +414,35 @@ def save_collection_association(reflection)
+      #
+      # In order to avoid this we temporarily disable autosave on associations
+      # leading back from child records to this parent.
+      def with_disabled_autosave_on_inverses(klass)
```
Is it necessary to pass the klass argument here? Is it safe to just use `self` instead?

egilburg commented on activerecord/lib/active_record/reflection.rb:
```diff
@@ -278,6 +277,14 @@ def autosave=(autosave)
         end
       end
 
+      def allow_autosave?
```
unused method

samphilipd replied:
Was going to use this in a refactor but... I'll leave that up to another PR. Better to keep this one strictly a bugfix.

arthurnn commented:
> This appears to have been a hack in order to avoid double-save problems with nested attributes

Do you have a link to a commit or PR about that?

Thanks for working on this.

samphilipd commented:
Here is the [guilty commit](https://github.com/rails/rails/commit/634fd040dbabf3e268a6a4bf4a9136468222022f).

I'm inferring this was a hack because this instance variable should definitely not be being set as a side-effect of this setter.

The commit message is not exactly helpful either.

You'd have to ask @tenderlove for the original story though.

samphilipd commented:
@arthurnn [here](26d19b4661f3d89a075b5f05d926c578ff0c730f) is the commit

matthewd commented:
Does this behave relatively intuitively if callbacks modify other records in the association-tree being saved?

The particular implementation is a non-starter because it completely ignores thread safety, but fixing that is easy enough.

@arthurnn this problem seems to have some conceptual overlap with deferred-touch.

samphilipd commented:
@matthewd how does it ignore threadsafety? Are the relation objects shared amongst threads?

samphilipd commented:
@matthewd Changed the implementation to one that is thread-safe (and probably more performant).

samphilipd commented:
@matthewd addressed your concern about thread-safety. The new implementation is quite a bit more readable as well.

Commit:
Fix falsely failing test reliant on ordering

- Without an explicit order by clause we cannot expect the database to
  return the records with any sort of order
- The ordering is incidental, in these tests we only care about the
  presence of the records

bf4 commented:
In https://github.com/rails/rails/pull/9522 @sgrif [commented](https://github.com/rails/rails/commit/26d19b4661f3d89a075b5f05d926c578ff0c730f#commitcomment-9482625)

> people don't expect accepts_nested_attributes_for to break automatic inverses.

and then

>  Turns out having autosave and inverse_of can cause a record to be saved more than once,
> because it's still marked as changed? inside of after_save, after_update, and after_create.
> And fixing that turned out to be an insurmountable task right now
> (assuming it's a breaking change we'd like to make which I'm not sure we do).

I have a patch of this against Rails 4-2 we're using

samphilipd commented:
@bf4 the double save problem is fixed in this pr.

sgrif commented:
I had meant to comment on this last week. This isn't the direction I'd like to go on master. This module, and those which interact with it is a giant ball of "flag to not double save in this one circumstance". I'd instead like to address the root cause of double save bugs in general, which is done in https://github.com/rails/rails/pull/25337 making this unnecessary.

However, we still need to fix https://github.com/rails/rails/issues/25198 for 5.0.1, where that patch cannot be applied. I think this is the most straightforward path to resolving there, but I want to explore one other alternative.

There's no need to close and re-open against another branch or anything like that. I will manually pull the commit over if we go forward with this solution for 5.0.1.

bf4 commented:
@samphilipd Thanks, [commented there](https://github.com/rails/rails/commit/26d19b4661f3d89a075b5f05d926c578ff0c730f#commitcomment-17826374) that I'm currently [using this patch applied to 4.2.6](https://gist.github.com/bf4/315a08632cfbd94c9fd33762e90457d5). I like how simple it is, though agree with @sgrif that it essentially duplicates existing behavior elsewhere. Working code > broken code :)

sgrif commented:
@samphilipd Do you mind re-opening this against 5-0-stable?
